src/clj/puppetlabs/puppetserver/ringutils.clj

```diff
@@ -119,4 +111,40 @@
       ; Our compojure app returns nil responses sometimes.
       ; In that case, don't add the header.
       (when response
-        (ring/header response "X-Puppet-Version" version)))))
\ No newline at end of file
+        (ring/header response "X-Puppet-Version" version)))))
+
+;; This function exists to support backward-compatible usage of a
+;; client-whitelist to protect access to some Clojure endpoints.  When support
+;; for client-whitelist authorization is dropped, this function should deleted.
+;; Callers would presumably be using a trapperkeeper-authorization handler
+;; for all endpoint authorization.
+(schema/defn ^:always-validate
+  wrap-with-trapperkeeper-or-client-whitelist-authorization
+  "Middleware function that routes a request through either an authorization
+  handler derived from the supplied 'authorization-fn' or to a client-whitelist
+  handler.
+
+  The 'authorization-fn' is expected to return a handler when called
+  and this function and accept a single argument, an downstream handler that the
+  authorization-fn should route a handled request to.  The authorization-fn
+  is called with 'base-handler' as its parameter.
+
+  Requests are only routed to the client-whitelist handler if the request 'uri'
+  starts with the value provided to this function for 'whitelist-path' and if
+  the 'whitelist-settings' are non-empty.  In all other cases, requests are
+  routed to the handler constructed from the 'authorization-fn'"
+  [base-handler :- IFn
+   authorization-fn :- IFn
+   whitelist-path :- schema/Str
+   whitelist-settings :- (schema/maybe WhitelistSettings)]
+  (let [handler-with-trapperkeeper-authorization (authorization-fn base-handler)]
+    (if-let [handler-with-client-whitelist-authorization
+             (if (not-empty whitelist-settings)
```

Comment:
It seems that using `(when)` when you don't have an else clause is more idiomatic, but this doesn't need to be fixed.